sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala

```diff
@@ -80,12 +81,21 @@ case class Filter(condition: Expression, child: SparkPlan)
   // Split out all the IsNotNulls from condition.
   private val (notNullPreds, otherPreds) = splitConjunctivePredicates(condition).partition {
     case IsNotNull(a) if child.output.contains(a) => true
+    case IsNotNull(a) =>
+      a match {
+        case Casts(a) if child.output.contains(a) => true
```

Comment:
We should not add these corner cases here, they should be handled by constraints.